File: AlgorithmsPackPythonSolutions/Minimum Number of Days to Make m Bouquets.py
class Solution:
    def possible(self, bloomDay, bouquet, flowers, day):
        adjFlowers = 0
        bouCount = 0
        
        for i in range(len(bloomDay)):
            if bloomDay[i] <= day:
                adjFlowers += 1
                
                if adjFlowers == flowers:
                    bouCount += 1
                    adjFlowers = 0
                if bouCount == bouquet:
                    return True
            else:
                adjFlowers = 0
        return False
    
    def minDays(self, bloomDay, m, k):
        sz = len(bloomDay)
        if m * k > sz:
            return -1
        
        start = 1
        end = int(1e9)
        ans = -1
        
        while start <= end:
            mid = start + (end - start) // 2
            if self.possible(bloomDay, m, k, mid):
                end = mid - 1
                ans = mid
            else:
                start = mid + 1
        return ans

File: AlgorithmsPackCPP/Minimum Number of Days to Make m Bouquets.cpp
class Solution {
public:
    bool possible(vector<int>&bloomDay, int bouquet, int flowers, int day){
        int adjFlowers = 0;
        int bouCount = 0;
        
        for(int i = 0; i < bloomDay.size(); i++){
            if(bloomDay[i]<=day){
                adjFlowers++;
                
                if(adjFlowers==flowers){
                    bouCount++;
                    adjFlowers = 0;
                }
                if(bouCount==bouquet)
                    return true;
            }else{
                adjFlowers = 0;
            }
        }
        return false;
    }
    
    int minDays(vector<int>& bloomDay, int m, int k) {
        int sz = bloomDay.size();
        if(m*k > sz)
            return -1;
        
        int start = 1;
        int end = 1e9;
        int ans = -1;
        
        while(start<=end){
            int mid = start+(end-start)/2;
            if(possible(bloomDay,m,k,mid)){
                end = mid-1;
                ans = mid;
            }else
                start = mid+1;
        }
        return ans;
    }
};
